parser/prism/Translator.cc

```diff
@@ -1,12 +1,46 @@
 #include "Translator.h"
+#include "ast/Helpers.h"
+#include "ast/Trees.h"
 
 template class std::unique_ptr<sorbet::parser::Node>;
 
-using std::make_unique;
 using std::unique_ptr;
 
 namespace sorbet::parser::Prism {
 
+using ast::MK;
+
+// A templated class which mimics the gvien `SorbetNode`, except with some extra room to fit an `ast::ExpressionPtr`.
+template <typename SorbetNode> class NodeAndExpr : public SorbetNode {
+    ast::ExpressionPtr desugaredExpr = nullptr;
+
+public:
+    template <typename... Args> NodeAndExpr(Args &&...args) : SorbetNode(std::forward<Args>(args)...) {}
+
+    virtual ast::ExpressionPtr getCachedDesugaredExpr() {
+        // We know each `NodeAndExpr` object's `getCachedDesugaredExpr()` will be called at most once, either:
+        // 1. When its parent node is being translated below, and this value is used to create that parent's expr.
+        // 2. When this node is visted by `node2TreeImpl` in `Runner.cc`, and this value is used in the fast-path.
+        //
+        // Because of this, we don't need to make any copies here. Just move this value out,
+        // and exclusive ownership to the caller.
+        return std::move(this->desugaredExpr);
+    }
+
+    // This method is intended to be called from the various `Translator` methods.
+    // Any value stored here will later be used to fast-path `node2TreeImpl` in `Desugar.cc`.
+    // This allows for a gradual migration to expression nodes, paving the way to eventually delete the translation
+    // to the intermediate `parser::Node` tree.
+    virtual void cacheDesugaredExpr(ast::ExpressionPtr desugaredExpr) {
+        this->desugaredExpr = std::move(desugaredExpr);
+    }
+};
+
+// Allocates a new `NodeAndExpr`, that's a subclass of `SorbetNode`, with no pre-computed `ExpressionPtr` AST.
+template <typename SorbetNode, typename... TArgs> unique_ptr<NodeAndExpr<SorbetNode>> make_node(TArgs &&...args) {
+    return std::make_unique<NodeAndExpr<SorbetNode>>(std::forward<TArgs>(args)...);
+}
```

Comment:
This helper doesn't pull much weight, but it saved me from needing to change all the call-sites below as I tweak all the details of exactly how `NodeAndExpr` works.